### PR TITLE
Social: Add the basic Open Graph tags

### PR DESCRIPTION
**Pull request by pablinos**
Without the `og:title` tag, the post doesn't pass validation for twitter and the card/featured image is not used. It also generates warnings on Facebook.

#### Changes proposed in this Pull Request:
This PR adds title, description and URL tags. It's mainly a copy of the code from `projects/plugins/jetpack/functions.opengraph.php`, which is a temporary measure until we refactor it to form a package.

#### Jetpack product discussion
p1HpG7-g6I-p2#comment-54679

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
- Create a site with Jetpack Social enabled and connected
- Publish a post with a featured image
- Put the URL into https://cards-dev.twitter.com/validator
- With this PR. Check that the card validates and that it displays as expected.

N.B. It can take two requests for the featured image to display.

### CODE REVIEW

**Comment by danielpost** [comment]
Nice! This is basically what I was doing as well. We should probably also copy over the [tests](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/tests/php/test_class.functions.opengraph.php) to the plugin.

**Comment by pablinos** [comment]
> We should probably also copy over the [tests](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/tests/php/test_class.functions.opengraph.php) to the plugin.

Yes, I've copied over the description tests in [4ef8924](https://github.com/Automattic/jetpack/pull/24547/commits/4ef892474247ff9e9049cfcccf2597fd7b314041), which seem the most relevant. We can add more tests when we move this to the package.

File: projects/plugins/social/src/class-meta-tags.php [hunk 1]
```diff
@@ -133,14 +134,14 @@ public function should_render_twitter_cards_tags() {
 	 * @param int $height  The minimum height of the image. Optional. Defaults to 200.
 	 * @return array The featured image and dimensions. Empty array if no image is found.
 	 */
-	public function get_featured_image( $post_id = 0, $width = 200, $height = 200 ) {
+	public function get_featured_image( $post_id = null, $width = 200, $height = 200 ) {
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:137]: Just curious, why this change?

File: projects/plugins/social/src/class-meta-tags.php [hunk 2]
```diff
@@ -133,14 +134,14 @@ public function should_render_twitter_cards_tags() {
...
 		$post = get_post( $post_id );
 
 		if ( ! has_post_thumbnail( $post ) ) {
 			return array();
 		}
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:142]: I just realised we might want to add an extra check here to see if `$post` is an actual post. Right now, you could pass in a non-existent post ID, and `$post` would return `null,` which would then make `has_post_thumbnail` check the global post object, rather than the post ID that was passed in.

File: projects/plugins/social/src/class-meta-tags.php [hunk 3]
```diff
@@ -169,34 +170,128 @@ public function get_featured_image( $post_id = 0, $width = 200, $height = 200 )
...
 	 */
-	public function render_tags() {
+	public function render_tags( $post = null ) {
 		if ( ! $this->should_render_meta_tags() ) {
 			return;
 		}
 
-		$image = $this->get_featured_image();
+		$data = get_post( $post );
+		$tags = array();
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:248]: Same here, we should probably do an early return if `$data === null`.

File: projects/plugins/social/src/class-meta-tags.php [hunk 4]
```diff
@@ -169,34 +170,128 @@ public function get_featured_image( $post_id = 0, $width = 200, $height = 200 )
...
+			$tags['og:title'] = wp_kses( apply_filters( 'the_title', $data->post_title, $data->ID ), array() );
 		}
 
-		$tags = array(
-			'og:image'        => $image['url'],
-			'og:image:width'  => $image['width'],
-			'og:image:height' => $image['height'],
-		);
+		$tags['og:url'] = get_permalink( $data->ID );
+		if ( ! post_password_required() ) {
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:257]: Should we pass in `$data->ID`?

File: projects/plugins/social/src/class-meta-tags.php [hunk 5]
```diff
@@ -169,34 +170,128 @@ public function get_featured_image( $post_id = 0, $width = 200, $height = 200 )
...
+		$data = get_post( $post );
+		$tags = array();
 
-		if ( empty( $image ) ) {
-			return;
+		if ( empty( $data->post_title ) ) {
+			$tags['og:title'] = __( '(no title)', 'jetpack-social' );
+		} else {
+			/** This filter is documented in core/src/wp-includes/post-template.php */
+			$tags['og:title'] = wp_kses( apply_filters( 'the_title', $data->post_title, $data->ID ), array() );
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:253]: I think this is just how it's set up in the Jetpack plugin right now, but I'm curious why we're doing a lot less sanitizing with the title compared to the description.

File: projects/plugins/social/src/class-meta-tags.php [hunk 6]
```diff
@@ -169,34 +170,128 @@ public function get_featured_image( $post_id = 0, $width = 200, $height = 200 )
...
+			 * If the post author set an excerpt, use that.
+			 * Otherwise, pick the post content that comes before the More tag if there is one.
+			 * Do not use the post content if it contains premium content.
+			 */
+			if ( ! empty( $data->post_excerpt ) ) {
+				$tags['og:description'] = $this->get_description( $data->post_excerpt );
+			} elseif ( ! has_block( 'premium-content/container', $data->post_content ) ) {
+				$excerpt                = explode( '<!--more-->', $data->post_content )[0];
+				$tags['og:description'] = $this->get_description( $excerpt );
+			}
```
Comment (danielpost) [inline projects/plugins/social/src/class-meta-tags.php:268]: Not for now, but when we extract this to a package we should make this filterable, so we don't need to check for specific blocks in the package.

**Review by danielpost — COMMENTED** [review]
I've added a few suggestions, but I think this is almost good to go!

File: projects/plugins/social/src/class-meta-tags.php (hunk 1 above)
Comment (pablinos) [reply projects/plugins/social/src/class-meta-tags.php:137]: It was for consistency as I thought all functions that accept a `WP_Post` or int default to `null` but as you pointed out, it's not very consistent in core and `get_the_title` defaults to `0`.

File: projects/plugins/social/src/class-meta-tags.php (hunk 4 above)
Comment (pablinos) [reply projects/plugins/social/src/class-meta-tags.php:257]: Yep, good catch. It wasn't necessary until we added in the $post parameter, but we should pass `$data` in. I think passing the object might prevent another call to `get_post`.

File: projects/plugins/social/src/class-meta-tags.php (hunk 5 above)
Comment (pablinos) [reply projects/plugins/social/src/class-meta-tags.php:253]: Yes, I noticed that too. I think part of it is that there is far less that could be in the title. Short codes etc, aren't relevant, and `wp_kses` should go a long way to stripping out anything we don't want.

I kind of feel that we should have a `sanitize_open_graph_text` function, that we could use for both.

File: projects/plugins/social/src/class-meta-tags.php (hunk 6 above)
Comment (pablinos) [reply projects/plugins/social/src/class-meta-tags.php:268]: Good plan. I think there's another filter later, that could be used for this. Definitely something we should do for the package.

File: projects/plugins/social/src/class-meta-tags.php (hunk 5 above)
Comment (danielpost) [reply projects/plugins/social/src/class-meta-tags.php:257]: > I kind of feel that we should have a sanitize_open_graph_text function, that we could use for both.

Agreed, we can do that for the package!

**Review by danielpost — APPROVED** [review]
